### PR TITLE
Hybrid retrieval tutorial

### PR DESCRIPTION
**Pull request by nickprock**
Added the Hybrid Retrieval Tutorial.

Issue #240

### CODE REVIEW

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "_gugk_k2lXRi"
+   },
+   "source": [
+    "### 4) Initialize JoinDocuments and Ranker\n",
+    "\n",
+    "While researching hybrid search, we needed a way to combine the results of BM25 and dense vector search into a single ranked list. It may not be obvious how to combine them:\n",
+    "\n",
+    "* Different retrievers use non-comparable score types, like BM25 and cosine similarity.\n",
```
Comment (dfokina) [inline tutorials/26_Hybrid_Retrieval.ipynb:273]:
```suggestion
    "* Different retrievers use incompatible score types, like BM25 and cosine similarity.\n",
```

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+    "* Documents may come from single or multiple sources at the same time. There should be a way to deal with duplicates in the final ranking.\n",
+    "\n",
+    "The merge and rank of the documents from different retrievers is an open problem, Haystack offers several methods in [`JoinDocuments`](https://docs.haystack.deepset.ai/docs/join_documents), here we use the simplest, `concatenate` and we postpone the task to the ranker.\n",
```
Comment (dfokina) [inline tutorials/26_Hybrid_Retrieval.ipynb:276]:
```suggestion
    "The merging and ranking of the documents from different retrievers is an open problem, however, Haystack offers several methods in [`JoinDocuments`](https://docs.haystack.deepset.ai/docs/join_documents). Here, you will use the simplest, `concatenate`, and pass the task to the ranker.\n",
```

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+    "\n",
+    "We use a [re-ranker based on a cross-encoder](https://docs.haystack.deepset.ai/docs/ranker#sentencetransformersranker) that scores the relevancy of all candidates for the given search query.\n",
```
Comment (dfokina) [inline tutorials/26_Hybrid_Retrieval.ipynb:278]:
```suggestion
    "Use a [re-ranker based on a cross-encoder](https://docs.haystack.deepset.ai/docs/ranker#sentencetransformersranker) that scores the relevancy of all candidates for the given search query.\n",
```

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+    "For more information about the `Ranker` check the [docs](https://docs.haystack.deepset.ai/docs/ranker)."
```
Comment (dfokina) [inline tutorials/26_Hybrid_Retrieval.ipynb:279]:
```suggestion
    "For more information about the `Ranker`, check the Haystack [docs](https://docs.haystack.deepset.ai/docs/ranker)."
```

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "PexSrsBLlXRp"
+   },
+   "source": [
+    "## 5) Create an Hybrid Retrieval Pipeline\n",
+    "\n",
+    "With a Haystack `Pipeline` you can stick together your building blocks to a search pipeline. Under the hood, `Pipelines` are Directed Acyclic Graphs (DAGs) that you can easily customize for your own use cases. To speed things up, Haystack also comes with a few predefined `Pipelines`.\n",
```
Comment (dfokina) [inline tutorials/26_Hybrid_Retrieval.ipynb:304]:
```suggestion
    "With a Haystack `Pipeline`, you can connect your building blocks into a search pipeline. Under the hood, `Pipelines` are Directed Acyclic Graphs (DAGs) that you can easily customize for your own use cases. To speed things up, Haystack also comes with a few predefined `Pipelines`.\n",
```

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+    "# pipeline.draw(\"pipeline_hybrid.png\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "sTTVLUJylXRx"
+   },
+   "source": [
+    "Search and article with Hybrid Retrieval, if you want to see all the steps you can enable `debug=True` in `JoinDocuments`'s `params`."
```
Comment (dfokina) [inline tutorials/26_Hybrid_Retrieval.ipynb:357]:
```suggestion
    "Search an article with Hybrid Retrieval. If you want to see all the steps, enable `debug=True` in `JoinDocuments`'s `params`."
```

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "WvPv1cJ6gbBJ"
+   },
+   "source": [
+    "Let's create a function to print a kind of *search page*."
```
Comment (dfokina) [inline tutorials/26_Hybrid_Retrieval.ipynb:386]:
```suggestion
    "Create a function to print a kind of *search page*."
```

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "rXxhhdQulXR5"
+   },
+   "source": [
+    "## About us\n",
+    "\n",
+    "*Leave this section as is. It's a footer that we add to all our tutorials.*\n",
```
Comment (dfokina) [inline tutorials/26_Hybrid_Retrieval.ipynb:423]: @bilgeyucel I'm not entirely sure if this line should stay here?

**Review by dfokina — CHANGES_REQUESTED** [review]
Thank you for the contribution @nickprock ! I left some language and grammar-related suggestions for the text.

File: tutorials/26_Hybrid_Retrieval.ipynb (hunk 8 above)
Comment (bilgeyucel) [reply tutorials/26_Hybrid_Retrieval.ipynb:423]: @nickprock you can just remove the "About Us" section completely

Comment (nickprock) [reply tutorials/26_Hybrid_Retrieval.ipynb:423]: Done

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+    "rerank = SentenceTransformersRanker(model_name_or_path=\"cross-encoder/ms-marco-MiniLM-L-6-v2\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "PexSrsBLlXRp"
+   },
+   "source": [
+    "## 5) Create an Hybrid Retrieval Pipeline\n",
```
Comment (bilgeyucel) [inline tutorials/26_Hybrid_Retrieval.ipynb:306]:
```suggestion
    "### 5) Create the Hybrid Retrieval Pipeline\n",
```

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+    "\n",
+    "With a Haystack `Pipeline`, you can connect your building blocks into a search pipeline. Under the hood, `Pipelines` are Directed Acyclic Graphs (DAGs) that you can easily customize for your own use cases. To speed things up, Haystack also comes with a few predefined `Pipelines`.\n",
```
Comment (bilgeyucel) [inline tutorials/26_Hybrid_Retrieval.ipynb:308]: No need to mention predefine pipelines as we dont use them in this tutorial:
```suggestion
    "With a Haystack `Pipeline`, you can connect your building blocks into a search pipeline. Under the hood, `Pipelines` are Directed Acyclic Graphs (DAGs) that you can easily customize for your own use cases.\n",
```

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+    "# pipeline.draw(\"pipeline_hybrid.png\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "sTTVLUJylXRx"
+   },
+   "source": [
+    "Search an article with Hybrid Retrieval. If you want to see all the steps, enable `debug=True` in `JoinDocuments`'s `params`."
```
Comment (bilgeyucel) [inline tutorials/26_Hybrid_Retrieval.ipynb:361]: You can add a new section titled as `## Trying Out the Hybrid Pipeline` here covering pipeline.run and printing parts

File: tutorials/26_Hybrid_Retrieval.ipynb [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "V3bsFkHuhHn4"
+   },
+   "source": [
+    "### Generating a Pipeline Diagram\n",
+    "\n",
+    "With any Pipeline, whether prebuilt or custom constructed, you can save a diagram showing how all the components are connected."
```
Comment (bilgeyucel) [inline tutorials/26_Hybrid_Retrieval.ipynb:337]: Now that we already have the pipeline image, why not use it?
```suggestion
    "With any Pipeline, whether prebuilt or custom constructed, you can save a diagram showing how all the components are connected. For example, the hybrid pipeline should look like this:"
```
And then you can add this :
```md
    ![Hybrid pipeline graph](https://github.com/deepset-ai/haystack-tutorials/raw/main/tutorials/img/tutorial_26_pipeline_hybrid.png)
```

File: tutorials/26_Hybrid_Retrieval.ipynb (hunk 11 above)
Comment (nickprock) [reply tutorials/26_Hybrid_Retrieval.ipynb:363]: done

**Review by bilgeyucel — APPROVED** [review]
Thank you @nickprock

**Review by dfokina — APPROVED** [review]

**Pull request merged**